lib.php

```diff
@@ -83,45 +86,62 @@ function tool_emailtemplate_pluginfile($course, $cm, $context, $filearea, $args,
 }
 
 /**
- * Stores the date when the footer image was last updated into a custom field.
+ * Stores tracking information about footer image version into the database.
  *
  * @param string $info
  * @return void
  */
-function tool_emailtemplate_update_timestamp($info) {
-    GLOBAL $CFG, $DB;
-
-    require_once($CFG->dirroot . '/user/profile/lib.php');
+function tool_emailtemplate_update_tracking($info) {
+    GLOBAL $DB;
 
-    $customfield = get_config('tool_emailtemplate', 'lastupdated');
-    $datelen = strlen(userdate(time(), get_string('dateformat', 'tool_emailtemplate')));
-    if (empty($customfield) || strlen($info) < ($datelen + 1) || !str_contains($info, '-')) {
+    // Confirm data is formatted correctly and contains the required info.
+    $date = date('Y-m-d');
+    $datelen = strlen($date);
+    if (strlen($info) < ($datelen + 1) || !str_contains($info, '-')) {
         return;
     }
 
-    $username = substr($info, 0, -$datelen - 1);
-    $date = substr($info, -$datelen);
-
     // Grab the user from the username. Doesn't handle cases where it's not unique.
+    $username = substr($info, 0, -$datelen - 1);
     $user = $DB->get_record('user', array('username' => $username));
     if (empty($user)) {
         return;
     }
 
-    profile_load_data($user);
+    // Verify that the remaining part of the string is a valid version id (date).
+    $version = strtotime(substr($info, -$datelen));
+    if (empty($version) || $version > time()) {
+        return;
+    }
+
+    // Use date timestamp rather than time so it's only updated once per day max.
+    $lastloaded = strtotime($date);
+
+    $updatedtimestamps = [
+        'version' => $version,
+        'lastloaded' => $lastloaded,
+    ];
 
-    $profilefieldname = 'profile_field_' . $customfield;
-    $currenttimestamp = $user->$profilefieldname;
+    // Load previous record for user.
+    $tracking = $DB->get_record('tool_emailtemplate_tracking', ['userid' => $user->id]);
 
-    try {
-        $date = new \DateTime($date);
-        $timestamp = $date->getTimestamp();
-        // Only save if later than the most recent saved timestamp.
-        if (isset($currenttimestamp) && $timestamp > $currenttimestamp) {
-            profile_save_custom_fields($user->id, [$customfield => $timestamp]);
+    // If we have no tracking information, create new record.
+    if (empty($tracking)) {
+        $updatedtimestamps['userid'] = $user->id;
+        $DB->insert_record('tool_emailtemplate_tracking', $updatedtimestamps);
+        return;
+    }
+
+    // Otherwise update existing record. Updates are checked individually to prevent overriding version info.
```

Comment:
this section feels a bit more complicated than expected, why loop over $updatedtimestamps?

Reply:
It was mostly to avoid duplication because both version and lastloaded need similar checks, but I can see how it would cause confusion, especially if more variables are added in the future. I'll get this updated.